### ISSUE TITLE
Implementing inaccessible squares

### ISSUE THREAD

**Issue by odedstr** [ianfab/Fairy-Stockfish#100]
I want to implement a way to define a variant with inaccessible squares, squares that no player can move to. I'm new to Fairy-Stockfish and trying to understand the code. 
I saw that if I enter a condition in Position::legal, and for example returns false for a whole rank, then it sometimes works as I expected and the engine never moves to this rank, but sometimes it does move to it, which results in the game ending with an error message "Black makes an illegal move" (testing with Cute Chess).
For example a condition like this causes this behavior:

```
Rank rank = rank_of(to);
if(rank == RANK_5)
{
    return false;
}
```

I'd appreciate any hints as to how to proceed to understand the code, and how I should go about implementing such a variant.

**Comment by ianfab** [comment]
If you have not seen it already, the [introduction to the code base](https://github.com/ianfab/Fairy-Stockfish/wiki/Understanding-the-code) in the wiki might help to get a first overview.

Would you just like to exclude pieces from moving to certain squares of the board or should those squares not even be part of the board (i.e., holes that can not be crossed)? For the former it is possible to exclude squares per piece type via the `mobilityRegion`, see e.g. the variant definition for xiangqi: https://github.com/ianfab/Fairy-Stockfish/blob/25549e2ee585ec5303a52640b37bedfdb338a461/src/variant.cpp#L822-L827
Boards with holes are not supported yet, see https://github.com/ianfab/Fairy-Stockfish/issues/53.

**Comment by odedstr** [comment]
I want to implement boards with holes.
Yes I read the introduction to the code base. Still, looking at the relevant code ( I assume movegen.cpp or position.cpp ) I'm a bit lost as to how to implement a new variant such as this.

**Comment by ianfab** [comment]
Implementing boards with holes is quite involved, since the attack bitboard generation needs to be changed to consider that holes can not be crossed, otherwise check detection can not work properly for sliders that are blocked by holes in the board. This is the main reason why I did not try to implement #53 yet.

What are the rules of the variant you would like to implement?

**Comment by odedstr** [comment]
It is for a game I'm planning with changing boards, sometimes rectangular boards, sometimes boards with barriers or holes. Other than that there will be a few more rule changes but only rules that are already supported by the variant configuration.

**Comment by ianfab** [comment]
I was asking because there might be easier workarounds depending on the rules of the game, otherwise it is hard to give very specific advice, and an implementation that works for all variants is quite difficult as mentioned before.
* Does it have royal pieces (kings)? If not, a broken attack bitboard generation might be acceptable, since it might cause bad performance, but could at least work.
* Does it have riders/sliders? If not, holes are equivalent to inaccessible squares and defining the `mobilityRegion` for all piece types might be sufficient.
* Does it have hoppers? If it has kings and sliders but at least no hoppers, than the attack bitboard generation could be faked by treating the holes as pieces on the board.

For a fully fledged implementation of boards with holes I currently see two options:
* Pass the bitboard of valid or invalid squares to `attacks_bb`/`moves_bb` to allow them to consider that. This might have quite a performance impact though.
* Re-initialize the magics when setting a variant depending on the set of accessible squares. This would not work for precomputed magics on older hardware though.
* Use a non-magic method for `rider_attacks_bb` and consider inaccessible squares in it.

**Comment by odedstr** [comment]
Yes it has a king, riders and hoppers.
I understand that it's quite difficult and a lot of work, but in any case thank you for the advice!

**Comment by ianfab** [comment]
Sure. And if you manage to get an implementation that might be useful to merge into this project, than a PR is very welcome (as contributions in general). For now I am closing this, since the basic questions seem to be clarified, but if you have further questions or comments, feel free to reopen it or continue discussion in #53 if applicable.

**Comment by odedstr** [comment]
Maybe it will be a simpler approach to support pieces that can't be captured? There is already support for immovable pieces, and if you can define a piece as both immovable and uncapturable it would effectively be a hole.

**Comment by ianfab** [comment]
That only works for sliders, but not for hoppers.

**Comment by odedstr** [comment]
Having the hopper be able to jump over the holes would actually be awesome for my game!

**Comment by ianfab** [comment]
Then they are not real holes any more, but yes in that case the workaround using immobile piece works. You can then just exclude those squares from the `mobilityRegion` of all pieces and put immobile pieces there in the starting position. This way you do not even need to implement anything, you can just define the variant using the existing options.

**Comment by odedstr** [comment]
Do I exclude the squares from the mobilityRegion using whiteDropRegion and blackDropRegion? I assumed this is referring to drops like in shogi and crazyhouse.

**Comment by ianfab** [comment]
No, `mobilityRegion` is currently not configurable via variants.ini, you need to add your variant to variant.cpp and recompile.

**Comment by odedstr** [comment]
Looks like it's working! One issue I found is that there isn't a stalemate declared if there are no more legal moves, but I'm not sure if it's a problem with engine or with how CuteChess works.

**Comment by pseudo13** [comment]
Hi, how do you exclude a field from the mobilityRegion?
I tried this, but it doesn't work.
`v->mobilityRegion[BLACK][Rook] = ~rank_bb(SQ_D5) & ~file_bb(SQ_D5);`
Thank you for your help.

**Comment by ianfab** [comment]
Unless there is a very specific reason why you need to change it in the code, I would recommend to define variants via the variants.ini file, also see the [wiki](https://github.com/ianfab/Fairy-Stockfish/wiki/Variant-configuration). Currently there is no way to specifically exclude squares, so you need to define all squares that are accessible, e.g., for all squares on an 8x8 board excluding d5:
`mobilityRegionBlackRook = *1 *2 *3 *4 a5 b5 c5 e5 f5 g5 h5 *6 *7 *8`
